lib/http2/connection.c

```diff
@@ -620,14 +622,40 @@ static int handle_incoming_request(h2o_http2_conn_t *conn, h2o_http2_stream_t *s
 
     h2o_probe_log_request(&stream->req, stream->stream_id);
 
-    int is_connect = h2o_memis(stream->req.input.method.base, stream->req.input.method.len, H2O_STRLIT("CONNECT"));
-
-    /* check existence of pseudo-headers */
-    int expected_map = H2O_HPACK_PARSE_HEADERS_METHOD_EXISTS | H2O_HPACK_PARSE_HEADERS_AUTHORITY_EXISTS;
-    if (!is_connect)
-        expected_map =
+    int is_connect, must_exist_map, may_exist_map;
+    if (h2o_memis(stream->req.input.method.base, stream->req.input.method.len, H2O_STRLIT("CONNECT"))) {
+        is_connect = 1;
+        must_exist_map = H2O_HPACK_PARSE_HEADERS_METHOD_EXISTS | H2O_HPACK_PARSE_HEADERS_AUTHORITY_EXISTS;
+        may_exist_map = 0;
+        /* extended connect looks like an ordinary request plus an upgrade token (:protocol) */
+        if ((header_exists_map & H2O_HPACK_PARSE_HEADERS_PROTOCOL_EXISTS) != 0)
+            must_exist_map |= H2O_HPACK_PARSE_HEADERS_SCHEME_EXISTS | H2O_HPACK_PARSE_HEADERS_PATH_EXISTS |
+                              H2O_HPACK_PARSE_HEADERS_PROTOCOL_EXISTS;
+    } else if (h2o_memis(stream->req.input.method.base, stream->req.input.method.len, H2O_STRLIT("CONNECT-UDP"))) {
+        /* Handling of masque draft-03. Method is CONNECT-UDP and :protocol is not used, so we set `:protocol` to "connect-udp" to
+         * make it look like an upgrade. The method is preserved and can be used to distinguish between RFC 9298 version which uses
+         * "CONNECT". */
+        if (!((header_exists_map & H2O_HPACK_PARSE_HEADERS_PROTOCOL_EXISTS) == 0 &&
+              stream->req.input.scheme == &H2O_URL_SCHEME_MASQUE &&
```

Comment:
PR #3281 proposes to relax checks on the `:scheme`.

We are currently handling traffic like this:

```
23529847 212 RxProtocol HTTP/3.0
23529847 212 RxHeader   :authority www.example.com:443
23529847 212 RxHeader   :method CONNECT
23529847 212 RxHeader   :scheme https
```

```
23535782 9 RxProtocol HTTP/2.0
23535782 9 RxHeader   :authority www.example.com:443
23535782 9 RxHeader   :method CONNECT
23535782 9 RxHeader   :scheme https
```

```
23529847 200 RxProtocol HTTP/3.0
23529847 200 RxHeader   :authority www.example.com:443
23529847 200 RxHeader   :method CONNECT-UDP
23529847 200 RxHeader   :path /
23529847 200 RxHeader   :scheme https
```

There are non-extended CONNECT requests with the pseudo-header field `:scheme` set to "https", even though the H2 and H3 specs require `:scheme` to be omitted in those requests.

There are draft-mode CONNECT-UDP requests with `:scheme` set to "https", even though the draft wants to see "masque" in `:scheme`.